<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>异常处理 | 余音的博客</title><meta name="keywords" content="博客，余音"><meta name="author" content="余音"><meta name="copyright" content="余音"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="异常的背景初识异常除以0 1234System.out.println(10 &#x2F; 0);&#x2F;&#x2F; 执行结果Exception in thread &quot;main&quot; java.lang.ArithmeticException: &#x2F; by zero    数组下标越界 12345int[] arr &#x3D; &amp;#123;1, 2, 3&amp;#125;;System.out.println(arr[1">
<meta property="og:type" content="article">
<meta property="og:title" content="异常处理">
<meta property="og:url" content="http://example.com/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/index.html">
<meta property="og:site_name" content="余音的博客">
<meta property="og:description" content="异常的背景初识异常除以0 1234System.out.println(10 &#x2F; 0);&#x2F;&#x2F; 执行结果Exception in thread &quot;main&quot; java.lang.ArithmeticException: &#x2F; by zero    数组下标越界 12345int[] arr &#x3D; &amp;#123;1, 2, 3&amp;#125;;System.out.println(arr[1">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif">
<meta property="article:published_time" content="2021-12-07T06:32:05.000Z">
<meta property="article:modified_time" content="2021-12-07T06:32:46.540Z">
<meta property="article:author" content="余音">
<meta property="article:tag" content="博客，余音">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"search.xml","languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery@2/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery@2/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '异常处理',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2021-12-07 14:32:46'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/constown/HexoCustomFile@0.0.4/dist/css/custom.min.css"><meta name="generator" content="Hexo 5.4.0"></head><body><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111251736783.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data is-center"><div class="data-item"><a href="/archives/"><div class="headline">文章</div><div class="length-num">10</div></a></div><div class="data-item"><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a></div><div class="data-item"><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 主页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fa fa-heartbeat"></i><span> 清单</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> 音乐</span></a></li><li><a class="site-page child" href="/Gallery/"><i class="fa-fw fas fa-images"></i><span> 照片</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> 电影</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友情链接</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">余音的博客</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 主页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fa fa-heartbeat"></i><span> 清单</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> 音乐</span></a></li><li><a class="site-page child" href="/Gallery/"><i class="fa-fw fas fa-images"></i><span> 照片</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> 电影</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友情链接</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">异常处理</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2021-12-07T06:32:05.000Z" title="发表于 2021-12-07 14:32:05">2021-12-07</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2021-12-07T06:32:46.540Z" title="更新于 2021-12-07 14:32:46">2021-12-07</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">字数总计:</span><span class="word-count">3.9k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">阅读时长:</span><span>15分钟</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="异常处理"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="异常的背景"><a href="#异常的背景" class="headerlink" title="异常的背景"></a>异常的背景</h1><h2 id="初识异常"><a href="#初识异常" class="headerlink" title="初识异常"></a>初识异常</h2><p><strong>除以0</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">System.out.println(<span class="number">10</span> / <span class="number">0</span>);</span><br><span class="line"></span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.ArithmeticException: / by zero</span><br></pre></td></tr></table></figure>



<p><strong>数组下标越界</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line">System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line"></span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.ArrayIndexOutOfBoundsException: <span class="number">100</span></span><br></pre></td></tr></table></figure>



<p><strong>访问null对象</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Test</span> </span>&#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">int</span> num = <span class="number">10</span>;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        Test t = <span class="keyword">null</span>;</span><br><span class="line">        System.out.println(t.num);</span><br><span class="line">   &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.NullPointerException</span><br></pre></td></tr></table></figure>

<p>所谓异常指的就是程序在 <strong>运行时</strong> 出现错误时通知调用者的一种机制</p>
<blockquote>
<p>关键字 “运行时”</p>
<p>有些错误是这样的, 例如将 <code>System.out.println</code> 拼写错了, 写成了 <code>system.out.println</code>. 此时编译过程中就会出错, 这是 “编译期” 出错.</p>
<p>而运行时指的是程序已经编译通过得到 class 文件了, 再由 JVM 执行过程中出现的错误</p>
</blockquote>
<p>异常的种类有很多, 不同种类的异常具有不同的含义, 也有不同的处理方式</p>
<p><strong>防御式编程</strong></p>
<p>错误在代码中是客观存在的. 因此我们要让程序出现问题的时候及时通知程序猿. 我们有两种主要的方式</p>
<p><strong>LBYL</strong>: Look Before You Leap. 在操作之前就做充分的检查.</p>
<p><strong>EAFP</strong>: It’s Easier to Ask Forgiveness than Permission. “事后获取原谅比事前获取许可更容易”. 也就是先操作, 遇到问题再处理.</p>
<p><strong>异常的核心思想就是 EAFP</strong></p>
<h2 id="异常的好处"><a href="#异常的好处" class="headerlink" title="异常的好处"></a>异常的好处</h2><p>例如, 我们用伪代码演示一下开始一局王者荣耀的过程.</p>
<p><strong>LBYL</strong> <strong>风格的代码</strong>(不使用异常)</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">boolean</span> ret = <span class="keyword">false</span>;</span><br><span class="line">ret = 登陆游戏();</span><br><span class="line"><span class="keyword">if</span> (!ret) &#123;</span><br><span class="line"> 处理登陆游戏错误;</span><br><span class="line">    <span class="keyword">return</span>; &#125;</span><br><span class="line">ret = 开始匹配();</span><br><span class="line"><span class="keyword">if</span> (!ret) &#123;</span><br><span class="line"> 处理匹配错误;</span><br><span class="line">    <span class="keyword">return</span>; &#125;</span><br><span class="line">ret = 游戏确认();</span><br><span class="line"><span class="keyword">if</span> (!ret) &#123;</span><br><span class="line"> 处理游戏确认错误;</span><br><span class="line">    <span class="keyword">return</span>; &#125;</span><br><span class="line">ret = 选择英雄();</span><br><span class="line"><span class="keyword">if</span> (!ret) &#123;</span><br><span class="line">    处理选择英雄错误;</span><br><span class="line">    <span class="keyword">return</span>; &#125;</span><br><span class="line">ret = 载入游戏画面();</span><br><span class="line"><span class="keyword">if</span> (!ret) &#123;</span><br><span class="line"> 处理载入游戏错误;</span><br><span class="line">    <span class="keyword">return</span>; &#125;</span><br><span class="line">......</span><br></pre></td></tr></table></figure>

<p><strong>EAFP</strong> <strong>风格的代码</strong>(使用异常)</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    登陆游戏();</span><br><span class="line">    开始匹配();</span><br><span class="line">    游戏确认();</span><br><span class="line">    选择英雄();</span><br><span class="line">    载入游戏画面();</span><br><span class="line">   ...</span><br><span class="line">&#125; <span class="keyword">catch</span> (登陆游戏异常) &#123;</span><br><span class="line">    处理登陆游戏异常;</span><br><span class="line">&#125; <span class="keyword">catch</span> (开始匹配异常) &#123;</span><br><span class="line"> 	处理开始匹配异常;</span><br><span class="line">&#125; <span class="keyword">catch</span> (游戏确认异常) &#123;</span><br><span class="line"> 	处理游戏确认异常;</span><br><span class="line">&#125; <span class="keyword">catch</span> (选择英雄异常) &#123;</span><br><span class="line"> 	处理选择英雄异常;</span><br><span class="line">&#125; <span class="keyword">catch</span> (载入游戏画面异常) &#123;</span><br><span class="line"> 	处理载入游戏画面异常; </span><br><span class="line">&#125;</span><br><span class="line">......</span><br></pre></td></tr></table></figure>

<p>对比两种不同风格的代码, 我们可以发现, 使用第一种方式, 正常流程和错误处理流程代码混在一起, 代码整体显的比较混乱. 而第二种方式正常流程和错误流程是分离开的, 更容易理解代码</p>
<h1 id="异常的基本用法"><a href="#异常的基本用法" class="headerlink" title="异常的基本用法"></a>异常的基本用法</h1><h2 id="捕获异常"><a href="#捕获异常" class="headerlink" title="捕获异常"></a><strong>捕获异常</strong></h2><p><strong>基本语法</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>&#123; </span><br><span class="line"> 有可能出现异常的语句 ; </span><br><span class="line">&#125;<span class="keyword">catch</span> (异常类型 异常对象) &#123;</span><br><span class="line">&#125; ...</span><br><span class="line">[<span class="keyword">finally</span> &#123;</span><br><span class="line"> 异常的出口</span><br><span class="line">&#125;]</span><br></pre></td></tr></table></figure>

<blockquote>
<p>try 代码块中放的是可能出现异常的代码.</p>
<p>catch 代码块中放的是出现异常后的处理行为.</p>
<p>finally 代码块中的代码用于处理善后工作, 会在最后执行.</p>
<p>其中 finally 都可以根据情况选择加或者不加.</p>
</blockquote>
<p><strong>代码示例1</strong> 不处理异常</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line">System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line"></span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.ArrayIndexOutOfBoundsException: <span class="number">100</span></span><br></pre></td></tr></table></figure>

<p>我们发现一旦出现异常, 程序就终止了. after 没有正确输出</p>
<p><strong>代码示例2</strong> 使用 try catch 后的程序执行过程</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">    System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line">&#125; <span class="keyword">catch</span> (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">    <span class="comment">// 打印出现异常的调用栈</span></span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br><span class="line">	System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">java.lang.ArrayIndexOutOfBoundsException: <span class="number">100</span></span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">10</span>)</span><br><span class="line">after <span class="keyword">try</span> <span class="keyword">catch</span></span><br></pre></td></tr></table></figure>

<p>我们发现, 一旦 try 中出现异常, 那么 try 代码块中的程序就不会继续执行, 而是交给 catch 中的代码来执行. catch 执行完毕会继续往下执行</p>
<blockquote>
<p><strong>关于异常的处理方式</strong></p>
<p>异常的种类有很多, 我们要根据不同的业务场景来决定.</p>
<p>对于比较严重的问题(例如和算钱相关的场景), 应该让程序直接崩溃, 防止造成更严重的后果</p>
<p>对于不太严重的问题(大多数场景), 可以记录错误日志, 并通过监控报警程序及时通知程序猿</p>
<p>对于可能会恢复的问题(和网络相关的场景), 可以尝试进行重试.</p>
<p>在我们当前的代码中采取的是经过简化的第二种方式. 我们记录的错误日志是出现异常的方法调用信息, 能很快速的让我们找到出现异常的位置. 以后在实际工作中我们会采取更完备的方式来记录异常信息</p>
</blockquote>
<blockquote>
<p><strong>关于</strong> <strong>“调用栈”</strong></p>
<p>方法之间是存在相互调用关系的, 这种调用关系我们可以用 “调用栈” 来描述. 在 JVM 中有一块内存空间称为 “虚拟机栈” 专门存储方法之间的调用关系. 当代码中出现异常的时候, 我们就可以使用     <code>e.printStackTrace();</code> 的方式查看出现异常代码的调用栈.</p>
</blockquote>
<p><strong>代码示例3</strong> catch 只能处理对应种类的异常</p>
<p>我们修改了代码, 让代码抛出的是空指针异常</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">    arr = <span class="keyword">null</span>;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">    System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line">&#125; <span class="keyword">catch</span> (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br><span class="line">System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.NullPointerException</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">11</span>)</span><br></pre></td></tr></table></figure>

<p>此时, catch 语句不能捕获到刚才的空指针异常. 因为异常类型不匹配</p>
<p><strong>代码示例4</strong> catch 可以有多个</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">    arr = <span class="keyword">null</span>;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">    System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line">&#125; <span class="keyword">catch</span> (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;这是个数组下标越界异常&quot;</span>);</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125; <span class="keyword">catch</span> (NullPointerException e) &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;这是个空指针异常&quot;</span>);</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br><span class="line">System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">这是个空指针异常</span><br><span class="line">java.lang.NullPointerException</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">12</span>)</span><br><span class="line">after <span class="keyword">try</span> <span class="keyword">catch</span></span><br></pre></td></tr></table></figure>

<p>一段代码可能会抛出多种不同的异常, 不同的异常有不同的处理方式. 因此可以搭配多个 catch 代码块.</p>
<p>如果多个异常的处理方式是完全相同, 也可以写成这样</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">catch</span> (ArrayIndexOutOfBoundsException | NullPointerException e) &#123;</span><br><span class="line"> ...</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>多个catch时，异常子类在上面，分类在下面</p>
</blockquote>
<p><strong>代码示例5</strong> 也可以用一个 catch 捕获所有异常(不推荐)</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">    arr = <span class="keyword">null</span>;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">    System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line">&#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br><span class="line">System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">java.lang.NullPointerException</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">12</span>)</span><br><span class="line">after <span class="keyword">try</span> <span class="keyword">catch</span></span><br></pre></td></tr></table></figure>

<p>由于 Exception 类是所有异常类的父类. 因此可以用这个类型表示捕捉所有异常.</p>
<blockquote>
<p>备注: catch 进行类型匹配的时候, 不光会匹配相同类型的异常对象, 也会捕捉目标异常类型的子类对象.</p>
<p>如刚才的代码, <code>NullPointerException</code> 和<code> ArrayIndexOutOfBoundsException</code> 都是 Exception 的子类, 因此都能被捕获到</p>
</blockquote>
<p><strong>代码示例6</strong> finally 表示最后的善后工作, 例如释放资源</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;before&quot;</span>);</span><br><span class="line">    arr = <span class="keyword">null</span>;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">    System.out.println(<span class="string">&quot;after&quot;</span>);</span><br><span class="line">&#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;finally code&quot;</span>);</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">before</span><br><span class="line">java.lang.NullPointerException</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">12</span>)</span><br><span class="line"><span class="keyword">finally</span> code</span><br></pre></td></tr></table></figure>

<p>无论是否存在异常, finally 中的代码一定都会执行到. 保证最终一定会执行到 Scanner 的 close 方法</p>
<p><strong>代码示例7</strong> 使用 try 负责回收资源</p>
<p>刚才的代码可以有一种等价写法, 将 Scanner 对象在 try 的 ( ) 中创建, 就能保证在 try 执行完毕后自动调用 Scanner的 close 方法.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> (Scanner sc = <span class="keyword">new</span> Scanner(System.in)) &#123;</span><br><span class="line">    <span class="keyword">int</span> num = sc.nextInt();</span><br><span class="line">    System.out.println(<span class="string">&quot;num = &quot;</span> + num);</span><br><span class="line">&#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>



<p><strong>代码示例8</strong> 如果本方法中没有合适的处理异常的方式, 就会沿着调用栈向上传递</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        func();</span><br><span class="line">   &#125; <span class="keyword">catch</span> (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">        e.printStackTrace();</span><br><span class="line">   &#125;</span><br><span class="line">    System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">func</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 直接结果</span></span><br><span class="line">java.lang.ArrayIndexOutOfBoundsException: <span class="number">100</span></span><br><span class="line"> at demo02.Test.func(Test.java:<span class="number">18</span>)</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">9</span>)</span><br><span class="line">after <span class="keyword">try</span> <span class="keyword">catch</span></span><br></pre></td></tr></table></figure>



<p><strong>代码示例9</strong> 如果向上一直传递都没有合适的方法处理异常, 最终就会交给 JVM 处理, 程序就会异常终止(和我们最开始未使用 try catch 时是一样的）</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">    func();</span><br><span class="line">        System.out.println(<span class="string">&quot;after try catch&quot;</span>);</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">func</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="keyword">int</span>[] arr = &#123;<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>&#125;;</span><br><span class="line">    System.out.println(arr[<span class="number">100</span>]);</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.ArrayIndexOutOfBoundsException: <span class="number">100</span></span><br><span class="line"> at demo02.Test.func(Test.java:<span class="number">14</span>)</span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">8</span>)</span><br></pre></td></tr></table></figure>

<p>可以看到, 程序已经异常终止了, 没有执行到 <code>System.out.println(&quot;after try catch&quot;);</code> 这一行</p>
<p><strong>异常处理流程</strong></p>
<blockquote>
<p>程序先执行 try 中的代码</p>
<p>如果 try 中的代码出现异常, 就会结束 try 中的代码, 看和 catch 中的异常类型是否匹配.</p>
<p>如果找到匹配的异常类型, 就会执行 catch 中的代码</p>
<p>如果没有找到匹配的异常类型, 就会将异常向上传递到上层调用者.</p>
<p>无论是否找到匹配的异常类型, finally 中的代码都会被执行到(在该方法结束之前执行).</p>
<p>如果上层调用者也没有处理的了异常, 就继续向上传递.</p>
<p>一直到 main 方法也没有合适的代码处理异常, 就会交给 JVM 来进行处理, 此时程序就会异常终止.</p>
</blockquote>
<p><strong>抛出异常</strong></p>
<p>除了 Java 内置的类会抛出一些异常之外, 程序猿也可以手动抛出某个异常. 使用 throw 关键字完成这个操作.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> 	System.out.println(divide(<span class="number">10</span>, <span class="number">0</span>)); </span><br><span class="line">&#125; </span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">int</span> <span class="title">divide</span><span class="params">(<span class="keyword">int</span> x, <span class="keyword">int</span> y)</span> </span>&#123; </span><br><span class="line"> 	<span class="keyword">if</span> (y == <span class="number">0</span>) &#123; </span><br><span class="line"> 		<span class="keyword">throw</span> <span class="keyword">new</span> ArithmeticException(<span class="string">&quot;抛出除 0 异常&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line"> 	<span class="keyword">return</span> x / y; </span><br><span class="line">&#125; </span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line">Exception in thread <span class="string">&quot;main&quot;</span> java.lang.ArithmeticException: 抛出除 <span class="number">0</span> 异常</span><br><span class="line"> at demo02.Test.divide(Test.java:<span class="number">14</span>) </span><br><span class="line"> at demo02.Test.main(Test.java:<span class="number">9</span>)</span><br></pre></td></tr></table></figure>

<p>在这个代码中, 我们可以根据实际情况来抛出需要的异常. 在构造异常对象同时可以指定一些描述性信息</p>
<p><strong>异常说明</strong></p>
<p>我们在处理异常的时候, 通常希望知道这段代码中究竟会出现哪些可能的异常.</p>
<p>我们可以使用 throws 关键字, 把可能抛出的异常显式的标注在方法定义的位置. 从而提醒调用者要注意捕获这些异常.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">int</span> <span class="title">divide</span><span class="params">(<span class="keyword">int</span> x, <span class="keyword">int</span> y)</span> <span class="keyword">throws</span> ArithmeticException </span>&#123; </span><br><span class="line"> 	<span class="keyword">if</span> (y == <span class="number">0</span>) &#123; </span><br><span class="line"> 		<span class="keyword">throw</span> <span class="keyword">new</span> ArithmeticException(<span class="string">&quot;抛出除 0 异常&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line"> 	<span class="keyword">return</span> x / y; </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><strong>关于</strong> <strong>finally</strong> <strong>的注意事项</strong></p>
<p>finally 中的代码保证一定会执行到. 这也会带来一些麻烦</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> 	System.out.println(func()); </span><br><span class="line">&#125; </span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">int</span> <span class="title">func</span><span class="params">()</span> </span>&#123; </span><br><span class="line"> 	<span class="keyword">try</span> &#123; </span><br><span class="line"> 		<span class="keyword">return</span> <span class="number">10</span>; </span><br><span class="line"> 	&#125; <span class="keyword">finally</span> &#123; </span><br><span class="line"> 		<span class="keyword">return</span> <span class="number">20</span>; </span><br><span class="line"> 	&#125; </span><br><span class="line">&#125; </span><br><span class="line"><span class="comment">// 执行结果</span></span><br><span class="line"><span class="number">20</span></span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>:</p>
<p>finally 执行的时机是在方法返回之前(try 或者 catch 中如果有 return 会在这个 return 之前执行 finally). 但是如果finally 中也存在 return 语句, 那么就会执行 finally 中的 return, 从而不会执行到 try 中原有的 return.</p>
<p>一般我们不建议在 finally 中写 return (被编译器当做一个警告).</p>
<h1 id="Java异常体系"><a href="#Java异常体系" class="headerlink" title="Java异常体系"></a>Java异常体系</h1><p>Java 内置了丰富的异常体系, 用来表示不同情况下的异常.</p>
<p>下图表示 Java 内置的异常类之间的继承关系:</p>
<blockquote>
<p>顶层类 Throwable 派生出两个重要的子类, Error 和 Exception</p>
<p>其中 Error 指的是 Java 运行时内部错误和资源耗尽错误. <strong>应用程序不抛出此类异常</strong>. 这种内部错误一旦出现，</p>
<p>除了告知用户并使程序终止之外, 再无能无力. 这种情况很少出现.</p>
<p>Exception 是我们程序猿所使用的异常类的父类.</p>
<p>其中 Exception 有一个子类称为 <code>RuntimeException</code> , 这里面又派生出很多我们常见的异常类</p>
<p><code>NullPointerException </code>, <code>IndexOutOfBoundsException</code> 等</p>
</blockquote>
<p>Java语言规范将派生于 Error 类或 <code>RuntimeException </code>类的所有异常称为 <strong>非受查异常</strong>, 所有的其他异常称为 <strong>受查异常</strong>.</p>
<p>如果一段代码可能抛出 <strong>受查异常</strong>, 那么必须显式进行处理.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> System.out.println(readFile()); </span><br><span class="line">&#125; </span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> String <span class="title">readFile</span><span class="params">()</span> </span>&#123; </span><br><span class="line"> <span class="comment">// 尝试打开文件, 并读其中的一行. </span></span><br><span class="line"> File file = <span class="keyword">new</span> File(<span class="string">&quot;d:/test.txt&quot;</span>); </span><br><span class="line"> <span class="comment">// 使用文件对象构造 Scanner 对象. </span></span><br><span class="line"> Scanner sc = <span class="keyword">new</span> Scanner(file); </span><br><span class="line"> <span class="keyword">return</span> sc.nextLine(); </span><br><span class="line">&#125; </span><br><span class="line"><span class="comment">// 编译出错</span></span><br><span class="line">Error:(<span class="number">13</span>, <span class="number">22</span>) java: 未报告的异常错误java.io.FileNotFoundException; 必须对其进行捕获或声明以便抛出</span><br></pre></td></tr></table></figure>

<p>查看 Scanner 的构造方法可以发现, 存在<code> FileNotFoundException</code> 这样的异常说明</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="title">Scanner</span><span class="params">(File source)</span> <span class="keyword">throws</span> FileNotFoundException </span>&#123; </span><br><span class="line"> ... </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>如 <code>FileNotFoundException</code> 这样的异常就是受查异常. 如果不显式处理, 编译无法通过.</p>
<p>显式处理的方式有两种:</p>
<p><strong>a)</strong> <strong>使用</strong> <strong>try catch</strong> <strong>包裹起来</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> System.out.println(readFile()); </span><br><span class="line">&#125; </span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> String <span class="title">readFile</span><span class="params">()</span> </span>&#123; </span><br><span class="line"> File file = <span class="keyword">new</span> File(<span class="string">&quot;d:/test.txt&quot;</span>); </span><br><span class="line"> Scanner sc = <span class="keyword">null</span>; </span><br><span class="line"> <span class="keyword">try</span> &#123; </span><br><span class="line"> sc = <span class="keyword">new</span> Scanner(file); </span><br><span class="line"> &#125; <span class="keyword">catch</span> (FileNotFoundException e) &#123; </span><br><span class="line"> e.printStackTrace(); </span><br><span class="line"> &#125; </span><br><span class="line"> <span class="keyword">return</span> sc.nextLine(); </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><strong>b)</strong> <strong>在方法上加上异常说明</strong>, <strong>相当于将处理动作交给上级调用者</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">public static void main(String[] args) &#123; </span><br><span class="line"> try &#123; </span><br><span class="line"> System.out.println(readFile()); </span><br><span class="line"> &#125; catch (FileNotFoundException e) &#123; </span><br><span class="line"> e.printStackTrace(); </span><br><span class="line"> &#125; </span><br><span class="line">&#125; </span><br><span class="line">public static String readFile() throws FileNotFoundException &#123; </span><br><span class="line"> File file = new File(&quot;d:/test.txt&quot;); </span><br><span class="line"> Scanner sc = new Scanner(file); </span><br><span class="line"> return sc.nextLine(); </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>别忘了 IDEA 神奇的 alt + enter, 能够快速修正代码</p>
</blockquote>
<h1 id="自定义异常类"><a href="#自定义异常类" class="headerlink" title="自定义异常类"></a>自定义异常类</h1><p> Java 中虽然已经内置了丰富的异常类, 但是我们实际场景中可能还有一些情况需要我们对异常类进行扩展, 创建符合我</p>
<p>们实际情况的异常.</p>
<p>例如, 我们实现一个用户登陆功能.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Test</span> </span>&#123; </span><br><span class="line"> 	<span class="keyword">private</span> <span class="keyword">static</span> String userName = <span class="string">&quot;admin&quot;</span>; </span><br><span class="line"> 	<span class="keyword">private</span> <span class="keyword">static</span> String password = <span class="string">&quot;123456&quot;</span>; </span><br><span class="line"> 	<span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> 		login(<span class="string">&quot;admin&quot;</span>, <span class="string">&quot;123456&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line"> 	<span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">login</span><span class="params">(String userName, String password)</span> </span>&#123; </span><br><span class="line"> 		<span class="keyword">if</span> (!Test.userName.equals(userName)) &#123; </span><br><span class="line"> 			<span class="comment">// TODO 处理用户名错误</span></span><br><span class="line"> 		&#125; </span><br><span class="line"> 		<span class="keyword">if</span> (!Test.password.equals(password)) &#123; </span><br><span class="line"> 			<span class="comment">// TODO 处理密码错误</span></span><br><span class="line"> 		&#125; </span><br><span class="line"> 		System.out.println(<span class="string">&quot;登陆成功&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>此时我们在处理用户名密码错误的时候可能就需要抛出两种异常. 我们可以基于已有的异常类进行扩展(继承), 创建和我们业务相关的异常类</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">UserError</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123; </span><br><span class="line"> 	<span class="function"><span class="keyword">public</span> <span class="title">UserError</span><span class="params">(String message)</span> </span>&#123; </span><br><span class="line"> 		<span class="keyword">super</span>(message); </span><br><span class="line"> 	&#125; </span><br><span class="line">&#125; </span><br><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">PasswordError</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123; </span><br><span class="line"> 	<span class="function"><span class="keyword">public</span> <span class="title">PasswordError</span><span class="params">(String message)</span> </span>&#123; </span><br><span class="line"> 		<span class="keyword">super</span>(message); </span><br><span class="line"> 	&#125; </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>此时我们的 login 代码可以改成</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123; </span><br><span class="line"> 	<span class="keyword">try</span> &#123; </span><br><span class="line"> 		login(<span class="string">&quot;admin&quot;</span>, <span class="string">&quot;123456&quot;</span>); </span><br><span class="line"> 	&#125; <span class="keyword">catch</span> (UserError userError) &#123; </span><br><span class="line"> 		userError.printStackTrace(); </span><br><span class="line"> 	&#125; <span class="keyword">catch</span> (PasswordError passwordError) &#123; </span><br><span class="line"> 		passwordError.printStackTrace(); </span><br><span class="line"> 	&#125; </span><br><span class="line">&#125; </span><br><span class="line"><span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">login</span><span class="params">(String userName, String password)</span> <span class="keyword">throws</span> UserError, </span></span><br><span class="line"><span class="function">PasswordError </span>&#123; </span><br><span class="line"> 	<span class="keyword">if</span> (!Test.userName.equals(userName)) &#123; </span><br><span class="line"> 		<span class="keyword">throw</span> <span class="keyword">new</span> UserError(<span class="string">&quot;用户名错误&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line"> 	<span class="keyword">if</span> (!Test.password.equals(password)) &#123; </span><br><span class="line"> 		<span class="keyword">throw</span> <span class="keyword">new</span> PasswordError(<span class="string">&quot;密码错误&quot;</span>); </span><br><span class="line"> 	&#125; </span><br><span class="line"> 	System.out.println(<span class="string">&quot;登陆成功&quot;</span>); </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><strong>注意事项</strong></p>
<blockquote>
<p>自定义异常通常会继承自 Exception 或者 <code>RuntimeException</code></p>
<p>继承自 Exception 的异常默认是受查异常</p>
<p>继承自 <code>RuntimeException</code> 的异常默认是非受查异常</p>
</blockquote>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="mailto:undefined">余音</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/">http://example.com/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="http://example.com" target="_blank">余音的博客</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/social-share.js/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/social-share.js/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2021/12/04/String%E7%B1%BB%E7%9A%84%E8%AE%A4%E8%AF%86/"><img class="prev-cover" src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252041482.jfif" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">String类的认识</div></div></a></div><div class="next-post pull-right"><a href="/2022/01/08/LeetCode%E6%95%B0%E7%BB%84%E5%88%B7%E9%A2%981/"><img class="next-cover" src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042639.jfif" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">LeetCode数组刷题1</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111251736783.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">余音</div><div class="author-info__description">余音的博客，欢迎访问</div></div><div class="card-info-data is-center"><div class="card-info-data-item"><a href="/archives/"><div class="headline">文章</div><div class="length-num">10</div></a></div><div class="card-info-data-item"><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a></div><div class="card-info-data-item"><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div></div><a class="button--animated" id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/fishyin" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="mailto:785811203@qq.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn card-announcement-animation"></i><span>公告</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E8%83%8C%E6%99%AF"><span class="toc-number">1.</span> <span class="toc-text">异常的背景</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%9D%E8%AF%86%E5%BC%82%E5%B8%B8"><span class="toc-number">1.1.</span> <span class="toc-text">初识异常</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E5%A5%BD%E5%A4%84"><span class="toc-number">1.2.</span> <span class="toc-text">异常的好处</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95"><span class="toc-number">2.</span> <span class="toc-text">异常的基本用法</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8D%95%E8%8E%B7%E5%BC%82%E5%B8%B8"><span class="toc-number">2.1.</span> <span class="toc-text">捕获异常</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Java%E5%BC%82%E5%B8%B8%E4%BD%93%E7%B3%BB"><span class="toc-number">3.</span> <span class="toc-text">Java异常体系</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8%E7%B1%BB"><span class="toc-number">4.</span> <span class="toc-text">自定义异常类</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2022/01/22/%E3%80%90Java%E9%9B%86%E5%90%88%E3%80%91ArrayList/" title="【Java集合】ArrayList"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042639.jfif" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【Java集合】ArrayList"/></a><div class="content"><a class="title" href="/2022/01/22/%E3%80%90Java%E9%9B%86%E5%90%88%E3%80%91ArrayList/" title="【Java集合】ArrayList">【Java集合】ArrayList</a><time datetime="2022-01-22T08:48:17.000Z" title="发表于 2022-01-22 16:48:17">2022-01-22</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2022/01/18/%E3%80%90Java%E9%9B%86%E5%90%88%E3%80%91%E6%8E%A5%E5%8F%A3%E7%9A%84%E7%AE%80%E5%8D%95%E4%BB%8B%E7%BB%8D/" title="【Java集合】接口的简单介绍"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【Java集合】接口的简单介绍"/></a><div class="content"><a class="title" href="/2022/01/18/%E3%80%90Java%E9%9B%86%E5%90%88%E3%80%91%E6%8E%A5%E5%8F%A3%E7%9A%84%E7%AE%80%E5%8D%95%E4%BB%8B%E7%BB%8D/" title="【Java集合】接口的简单介绍">【Java集合】接口的简单介绍</a><time datetime="2022-01-18T08:34:46.000Z" title="发表于 2022-01-18 16:34:46">2022-01-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2022/01/08/LeetCode%E6%95%B0%E7%BB%84%E5%88%B7%E9%A2%981/" title="LeetCode数组刷题1"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042639.jfif" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="LeetCode数组刷题1"/></a><div class="content"><a class="title" href="/2022/01/08/LeetCode%E6%95%B0%E7%BB%84%E5%88%B7%E9%A2%981/" title="LeetCode数组刷题1">LeetCode数组刷题1</a><time datetime="2022-01-08T05:43:44.000Z" title="发表于 2022-01-08 13:43:44">2022-01-08</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/" title="异常处理"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252042119.jfif" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="异常处理"/></a><div class="content"><a class="title" href="/2021/12/07/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/" title="异常处理">异常处理</a><time datetime="2021-12-07T06:32:05.000Z" title="发表于 2021-12-07 14:32:05">2021-12-07</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/12/04/String%E7%B1%BB%E7%9A%84%E8%AE%A4%E8%AF%86/" title="String类的认识"><img src="https://cdn.jsdelivr.net/gh/fishyin/picodemo/img/202111252041482.jfif" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="String类的认识"/></a><div class="content"><a class="title" href="/2021/12/04/String%E7%B1%BB%E7%9A%84%E8%AE%A4%E8%AF%86/" title="String类的认识">String类的认识</a><time datetime="2021-12-04T02:06:00.000Z" title="发表于 2021-12-04 10:06:00">2021-12-04</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2021 - 2022 By 余音</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Thank you for visiting my blog！</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">本地搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.js"></script><script src="/js/search/local-search.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>